polkadot/node/core/approval-voting-parallel/src/lib.rs

```diff
@@ -0,0 +1,950 @@
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The Approval Voting Parallel Subsystem.
+//!
+//! This subsystem is responsible for orchestrating the work done by
+//! approval-voting and approval-distribution subsystem, so they can
+//! do their work in parallel, rather than serially, when they are run
+//! as independent subsystems.
+use itertools::Itertools;
+use metrics::{Meters, MetricsWatcher};
+use polkadot_node_core_approval_voting::{Config, RealAssignmentCriteria};
+use polkadot_node_metrics::metered::{
+	self, channel, unbounded, MeteredReceiver, MeteredSender, UnboundedMeteredReceiver,
+	UnboundedMeteredSender,
+};
+
+use polkadot_node_primitives::{
+	approval::time::{Clock, SystemClock},
+	DISPUTE_WINDOW,
+};
+use polkadot_node_subsystem::{
+	messages::{ApprovalDistributionMessage, ApprovalVotingMessage, ApprovalVotingParallelMessage},
+	overseer, FromOrchestra, SpawnedSubsystem, SubsystemError, SubsystemResult,
+};
+use polkadot_node_subsystem_util::{
+	self,
+	database::Database,
+	runtime::{Config as RuntimeInfoConfig, RuntimeInfo},
+};
+use polkadot_overseer::{OverseerSignal, Priority, SubsystemSender, TimeoutExt};
+use polkadot_primitives::{CandidateIndex, Hash, ValidatorIndex, ValidatorSignature};
+use rand::SeedableRng;
+
+use sc_keystore::LocalKeystore;
+use sp_consensus::SyncOracle;
+
+use futures::{channel::oneshot, prelude::*, StreamExt};
+pub use metrics::Metrics;
+use polkadot_node_core_approval_voting::{
+	approval_db::common::Config as DatabaseConfig, ApprovalVotingWorkProvider,
+};
+use std::{
+	collections::{HashMap, HashSet},
+	fmt::Debug,
+	sync::Arc,
+	time::Duration,
+};
+use stream::{select_with_strategy, PollNext, SelectWithStrategy};
+pub mod metrics;
+
+#[cfg(test)]
+mod tests;
+
+pub(crate) const LOG_TARGET: &str = "parachain::approval-voting-parallel";
+// Value rather arbitrarily: Should not be hit in practice, it exists to more easily diagnose dead
+// lock issues for example.
+const WAIT_FOR_SIGS_GATHER_TIMEOUT: Duration = Duration::from_millis(2000);
+
+/// The approval voting subsystem.
+pub struct ApprovalVotingParallelSubsystem {
+	/// `LocalKeystore` is needed for assignment keys, but not necessarily approval keys.
+	///
+	/// We do a lot of VRF signing and need the keys to have low latency.
+	keystore: Arc<LocalKeystore>,
+	db_config: DatabaseConfig,
+	slot_duration_millis: u64,
+	db: Arc<dyn Database>,
+	sync_oracle: Box<dyn SyncOracle + Send>,
+	metrics: Metrics,
+	spawner: Arc<dyn overseer::gen::Spawner + 'static>,
+	clock: Arc<dyn Clock + Send + Sync>,
+	subsystem_enabled: bool,
```

Comment:
The reason I went with subsystem_disabled for approval-distribution and approval-voting and subsystem_enabled in approval-voting-parallel, is because you can pass straight on the approval_voting_parallel_enabled flag as arguments without having to pay attention where you have to negate and where you don't have do it.

But, I see no strong reasons you can't do it as you suggest :D.

Reply:
Done.